[WIP] add optional sized int list support in schemas

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36867 scratch changes for optional int list testing
* **#36866 [WIP] add optional sized int list support in schemas**

E.g. `foo(int[]? x)`, `foo(int[3]? x)`

Commit:
[WIP] add optional sized int list support in schemas

E.g. `foo(int[3]? x)`

Basic plumbing is in place, couple obvious bugs but out of time for now

* vector<int64_t> to IntArrayRef lifetime issues
* fixed-size defaulting behavior is broken somewhere

[ghstack-poisoned]

Commit:
Update on "[WIP] add optional sized int list support in schemas"


E.g. `foo(int[3]? x)`

Basic plumbing is in place, couple obvious bugs but out of time for now

* vector<int64_t> to IntArrayRef lifetime issues
* fixed-size defaulting behavior is broken somewhere

[ghstack-poisoned]

Commit:
Update on "[WIP] add optional sized int list support in schemas"


E.g. `foo(int[3]? x)`

Basic plumbing is in place, couple obvious bugs but out of time for now

* vector<int64_t> to IntArrayRef lifetime issues
* fixed-size defaulting behavior is broken somewhere

[ghstack-poisoned]

dreiss commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -684,6 +684,13 @@ inline std::vector<int64_t> IValue::toIntVector() const {
+  if (isNone()) {
+    return c10::nullopt;
+  }
+  return at::IntArrayRef(toIntVector());
```
This isn't safe. The vector will be destroyed as this function returns, and we'll return an IntArrayRef that points to freed memory.

bhosmer replied:
Yup, haven't updated the PR description but I'm working through these obvious ones now.

Commit:
Update on "[WIP] add optional sized int list support in schemas"


E.g. `foo(int[3]? x)`

Eager side is working, JIT side has (at least) a std::vector<int64_t> => IntArrayRef lifetime issue somewhere.

[ghstack-poisoned]

Commit:
Update on "[WIP] add optional sized int list support in schemas"


E.g. `foo(int[3]? x)`

Eager side is working, JIT side has (at least) a std::vector<int64_t> => IntArrayRef lifetime issue somewhere.

[ghstack-poisoned]

bhosmer commented:
@dreiss pushing the latest version of the patch for reference - has all the necessary schema parsing for eager and JIT (including the definite size option, but that's easily omitted/suppressed), along with codegen changes to guarantee lifetimes of IntArrayRef-wrapped std::vectors.

#36867 is a throwaway patch that 
* mods the variants of `upsample_linear1d` in native_functions.yaml to take `int[1]? output_size_opt`
* mods the kernels to take c10::optional<IntArrayRef> instead of IntArrayRef in that position (works like the original when a list is passed, otherwise prints and exits)

Tested against the following, works as expected
```
import torch

t = torch.tensor(range(3 * 4 * 5), dtype=torch.float).reshape(3, 4, 5)

def foo(t):
  return torch._C._nn.upsample_linear1d(t, [1], True)

print("HEY eager")
print(foo(t))

print("HEY traced")
traced = torch.jit.trace(foo, t)
print(traced(t))

print("HEY scripted")
scripted = torch.jit.script(foo)
print(scripted(t))
```
